### DISCUSSION TITLE
Displaying item properties on the InventoryGrid

### DISCUSSION THREAD

**Discussion by bensgameprojects** [peter-kish/gloot#74] [Q&A] [answered]
Hi I'm trying to use InventoryGrid for my inventory system. My item protoset looks like this:
```
	{
		"id": "Blueberry",
		"width": 1,
		"height":1,
		"stack_size":1,
		"max_stack_size": 999,
		"is_stackable":true,
		"image": "res://Items/Fruits Asset/Black Outline/01.png"
	},
[ctrl_inventory_item_rect.zip](https://github.com/peter-kish/gloot/files/10110976/ctrl_inventory_item_rect.zip)

```
I want to display the stack_size using a RichTextLabel or similar positioned on the corner of the items in the CtrlInventoryGrid. I thought to add the modifications to ctrl_inventory_item_rect.gd. I added a new RichTextLabel var that gets instanced when the script is called and it is set and displayed in the _draw() function. Can someone please point me in the right direction? Thanks in advance. Attached is my changes to ctrl_inventory_item_rect.gd (that doesn't work).
[ctrl_inventory_item_rect.zip](https://github.com/peter-kish/gloot/files/10110976/ctrl_inventory_item_rect.zip)

**Comment by peter-kish** [comment]
I gave it a quick try and, strangely, replacing `RichTextLabel` with a `Label` seems to work and I don't really know why.

But you probably don't need to call `_update_stack_size_label()` inside `_draw()`.
What you probably want to do is update the label when item properties change. There is a `InventoryItem` signal that is triggered in that case: `properties_changed`.
I think calling `_update_stack_size_label()` inside `_refresh` and on `properties_changed` should be enough. Just make sure to disconnect/reconnect the signal when a different item is set (i.e. inside `_set_item(new_item)`).

I'll play around with it a bit more to figure out why `RichTextLabel` behaves differently from `Label`.

**Reply by bensgameprojects** [reply]
Thank you very much! I followed what you suggested and it seems to work. I call _update_stack_size_label() inside _refresh() but I didn't add a call on properties_changed. I can see that moving the item from one slot to another causes _update_stack_size_label to be called 4 times already with the only call being inside _refresh().
 
Could you elaborate a little bit on what you mean by disconnect/reconnect the signal inside _set_item(new_item)? I am handling the merges as an alternate case in inventory_grid.gd : move_item_to() when rect_free returns false and there's another item there that has the same prototype_id. Is that the appropriate place or should it be handled in _set_item()? Below is the code for my modified move_item_to method with stack size handling. As a side note is emit_signal("item_removed", item) redundant after using the remove_item method? I'm still trying to wrap my head around the organization and signalling processes in gloot.

Again, thank you!!

```
func move_item_to(item: InventoryItem, position: Vector2) -> bool:
	var item_size = get_item_size(item)
	var rect: Rect2 = Rect2(position, item_size)
	var other_item = get_item_at(position)
	if rect_free(rect, item):
		item.properties[KEY_GRID_POSITION] = position
		if item.properties[KEY_GRID_POSITION] == Vector2.ZERO:
			item.properties.erase(KEY_GRID_POSITION)
		emit_signal("contents_changed")
		return true
	# we can stack items here by checking
	# in the case !rect_free(rect, item) &&
	# get_item_at(position).get_property("id") == item.get_property("id")
	# compare that they are the same item id
	# also ensure the items are stackable type
	elif other_item != null && other_item.prototype_id == item.prototype_id && item.get_property("is_stackable"):
		var item_stack_size = item.get_property("stack_size")
		var other_item_stack_size = other_item.get_property("stack_size")
		var item_max_stack_size = item.get_property("max_stack_size")
		var new_other_item_stack_size = item_stack_size + other_item_stack_size
		var new_item_stack_size = 0
		# logic to handle max stack size merges
		if(new_other_item_stack_size > item_max_stack_size):
			new_item_stack_size = new_other_item_stack_size - item_max_stack_size
			item.set_property("stack_size", new_item_stack_size)
			new_other_item_stack_size -= new_item_stack_size
		else:
			# the item was completely merged
			# remove the donating item from the inventory (and existence :o)
			self.remove_item(item)
			emit_signal("item_removed", item)
		
		# now set the receiving items new stack_size
		other_item.set_property("stack_size", new_other_item_stack_size)
		return true
		
	return false
```

**Reply by peter-kish** [reply]
> I call _update_stack_size_label() inside _refresh() but I didn't add a call on properties_changed.

In that case calling `_update_stack_size_label()` on `properties_changed` might not be needed after all. I was just worried that in some cases the label might not get updated to reflect the item properties accordingly. But you can always add that later if you notice that kind of issues.

> I can see that moving the item from one slot to another causes _update_stack_size_label to be called 4 times already with the only call being inside _refresh()

Yeah, the `Ctrl*` classes tend to refresh more often than necessary in the current implementation. I hope to optimize that in the future.

> Could you elaborate a little bit on what you mean by disconnect/reconnect the signal inside _set_item(new_item)?

This just means that if you are "subscribed" to the `properties_changed` signal of the referenced item, you probably want to "unsubscribe" when the referenced item changes and the `CtrlInventoryItemRect` doesn't represent the old item anymore (and "subscribe" to the `properties_changed` signal of the newly referenced item). And the best place to do that is inside `_set_item()` which is called whenever the `item` field changes.
Hope that makes sense :)

This is another thing that might not be completely necessary, as I don't think we are ever doing this dynamic changing of the referenced item in `CtrlInventoryItemRect`.

> I am handling the merges as an alternate case in inventory_grid.gd : move_item_to() when rect_free returns false and there's another item there that has the same prototype_id. Is that the appropriate place or should it be handled in _set_item()?

Yeah, doing that in `move_item_to()` should be fine.

> As a side note is emit_signal("item_removed", item) redundant after using the remove_item method?

Yup, it's redundant. `item_removed` will be emitted when `remove_item()` is called.
It's not very obvious by looking at the code, but it is emitted in `InventoryItem._emit_removed()` when the item node gets unparented.

**Comment by peter-kish** [comment, marked as answer]
Ok, it seems like you need to set the `rect_min_size` for the `RichTextLabel` as suggested [here](https://godotengine.org/qa/15546/richtextlabel-does-not-show-in-gridcontainer).

Adding the following line to `_update_stack_size_label()` should do the trick:
```
stack_size_label.rect_min_size = rect_size
```
